### ISSUE TITLE
Saving optimizer state

### ISSUE THREAD

**Issue by huzecong** [clab/dynet#785]
Currently there are no methods to save the state of the optimizer, for instance current learning rate or history of gradients. These may be useful when resuming training from a model snapshot.

Could we have an API for this? I think a nice interface would be (for Python) `trainer.populate` after creating the Trainer object given the Model.

**Comment by neubig** [comment]
@xunzhang I think this shouldn't be too hard. What do you think?

**Comment by alsuhr-c** [comment]
To clarify, what is the current behavior of Dynet when it comes to snapshots? If I have an Adam optimizer for example, are the rates associated with each parameter saved and re-loaded into the model when I populate the parameters? Or, are these rates reset to the default values (this post would indicate this is true)?

**Comment by FilippoC** [comment]
Hi,
is there any update on this? Is there any way to save and reload the state of an optimizer?
It would be nice if we could save/load the parameters of the optimizer without having to export it to a file, i.e. in an object.

**Comment by philip30** [comment]
Hi, is there any python binding for saving the optimizer state?
